File: fixtures/scale/scala/medium/src/calc/funcs/Stats.scala
package calc.funcs

import calc.eval.FunctionRegistry

object Stats {
  def register(registry: FunctionRegistry): Unit = {
    registry.addAggregate("min", _.min)
    registry.addAggregate("max", _.max)
    registry.addAggregate("sum", _.sum)
    registry.addAggregate("mean", mean)
    registry.addAggregate("stddev", stddev)
  }

  private[funcs] def mean(args: Array[Double]): Double = args.sum / args.length

  private[funcs] def stddev(args: Array[Double]): Double = {
    val m = mean(args)
    val sumSquares = args.map(x => (x - m) * (x - m)).sum
    math.sqrt(sumSquares / math.max(1, args.length - 1))
  }
}

File: fixtures/scale/kotlin/medium/src/calc/funcs/Stats.kt
package calc.funcs

import calc.eval.FunctionRegistry
import kotlin.math.sqrt

fun registerStats(registry: FunctionRegistry) {
    registry.addAggregate("min") { it.min() }
    registry.addAggregate("max") { it.max() }
    registry.addAggregate("sum") { it.sum() }
    registry.addAggregate("mean") { mean(it) }
    registry.addAggregate("stddev") { stddev(it) }
}

internal fun mean(args: DoubleArray): Double = args.average()

internal fun stddev(args: DoubleArray): Double {
    val m = mean(args)
    val sum = args.sumOf { (it - m) * (it - m) }
    return sqrt(sum / (args.size - 1).coerceAtLeast(1))
}
